proposals/0000-create-asset-instruction.md

```diff
@@ -28,10 +28,23 @@ The instruction would have the following specification:
 CreateAsset {asset: MultiAsset, owner: MultiLocation}
 ```
 
-The `CreateAsset` instruction allows for the creation of a single asset. This can be a fungible or non fungible asset.
+The `CreateAsset` instruction allows for the creation of a single asset. This can be a fungible or non-fungible asset based on the specification of the `asset` field. The MultiAsset can describe a single asset. The `id` field is either an asset identity for a fungible asset or a class for a non-fungible asset. The fun field represents either the amount in the case of a fungible asset or the instance Id for a non-fungible asset. 
```

Comment:
These instructions are not connected to FRAME-based systems, so this instruction should also be usable to create an ERC20 token, for example, where the amount can represent the total_supply. 

Like you said, the amount could represent the min_balance for the assets_pallet. 

I do not directly see the difference between the CreateAsset with the assets field or with the id and fun fields split as the MultiAsset has both of these fields. However, if we split the fields, we could set the `fun` to `Option<Fungibility>`.

Reply:
thx, I know they are not connected and xcm should be disconnected from frame as much possible,

ideally, fields should be self explanatory and also to avoid confusion or misuse, 
so e.g. `MultiAsset.fun.amount` if user thinks about this amount as "total_supply" and on target system, we use this with several "AssetTransactors" where one can interpret this as "total_supply" and other as "min_balance", it could lead to problems
so for this case, fields split, as you said with Option could work better:
```
CreateAsset {
    id: Concrete(MultiLocation {parents: 1, interior: X1(GeneralIndex(1))})
    total_supply: Option<Funbility>,
    min_balance: Option<Funbility>,
    owner: MultiLocation {parents: 1, interior: Here},
}
```
we can go also with some properties wrapper and so on :)
```
AssetProperties {
    total_supply: Option<Funbility>,
    min_balance: Option<Funbility>,
}
```

yeah, I understand that this should be pretty generic  to cover all cases,
so I am interesting where this will go :)

Reply:
Yeah you are definitely right that it could cause serious problems (imagining an asset that has it's min balance set to value that should be total_supply haha). I would like to see more feedback on this approach. Especially the AssetProperties approach as it might also solve the Metadata related question.

Reply:
I like the idea of an `AssetProperties`, but I think we can't reuse `Fungibility` for those fields.
I think a possible approach would be to make the fungibility separation in the properties themselves.

```rust
enum AssetProperties {
  Fungible {
    total_supply: u128,
    min_balance: u128,
    // probably some other fields
  },
  NonFungible {
    collection_id: u64,
    // probably more fields
  },
}
```

If we add the largest amount of properties pertaining to that fungibility as possible, if some implementations don't handle some of them, like collections for example, they can just ignore that field.

Reply:
We could reasonably have an `AssetProperties` field in there I guess. It'll be a bit weird through since we'll need to make all field `Option`al otherwise we'll force the sender to introduce garbage info for fields which don't make sense for their asset and hope that the receiver ignores them. But then not all receive will support all fields. I guess we can have a convention of strictness, whereby receivers require some fields to be `Some` (if their implementation really needs them) or `None` (if their implementation doesn't support them and they're actually important). Other fields it might be flexible on.

```
enum Instruction {
  /* snip */
  InitializeAsset {
    /// The (concrete) asset ID, relative to the receiver/executor. This is what can
    /// be used as an asset ID when minting or teleporting assets into the chain.
    asset_id: MultiLocation,
    /// The owner location, relative to the receiver/executor. Some implementations
    /// might support all kinds of locations, others might only support
    /// `{ parents: 0, interior: X1(AccountId32 { .. }) }` patterns.
    owner: MultiLocation,
    /// Relevant properties of the asset class/ID to be initialized. The
    /// implementation may require some fields to be `Some` and others `None`.
    properties: AssetProperties,
  }
}
enum AssetProperties {
  Fungible {
    total_supply: Option<u128>,
    min_balance: Option<u128>,
  },
  NonFungible,
}
```

What would the proposed `collection_id` be? We have a named `owner` already for the asset class/ID who can actually mint the (NF)Tokens. If `collection_id` is passed anywhere, I'd have expected it to be passed when minting, not initializing.

Reply:
I like the idea of calling the instruction `InitializeAsset`, as well as adding the AssetProperties field. Is there already a standard for NFT metadata?